### PR TITLE
gui: Drop boost::scoped_array and use wchar_t API explicitly on Windows

### LINKED ISSUE #13819: Windows: Can not create startup link if the username is a_ä_α_Б
When selecting "Start Bitcoin Core on system login" in the settings, it will unselect itself after closing the dialog.

Tested with current master (e8ffec6) on Windows 7.

### PR DESCRIPTION
**Pull request by ken2812221**
Drop boost::scoped_array and simplify the code.

`TCHAR` should be defined as `wchar_t` if `UNICODE` is defined. So we can use `.toStdWString().c_str()` to get wchar_t C-style string.

Fix #13819

### CODE REVIEW

**Comment by fanquake** [comment]
This was introduced in #5793. Looks like something similar to what you are doing was proposed at the time as well.

**Comment by ken2812221** [comment]
Hope there is a way to see code that is before he force push.

**Comment by practicalswift** [comment]
Concept ACK

Thanks for removing Boost dependencies. Keep it coming! :-)

**Comment by MarcoFalke** [comment]
Any suggestion on how to test this?

**Comment by ken2812221** [comment]
This is Windows-only code, can be tested as same as https://github.com/bitcoin/bitcoin/pull/5793#issuecomment-77312225

File: src/qt/guiutil.cpp [hunk 1]
```diff
@@ -625,7 +615,7 @@ bool SetStartOnSystemStartup(bool fAutoStart)
 #ifndef UNICODE
             psl->SetArguments(strArgs.toStdString().c_str());
 #else
```
Comment (Sjors) [inline src/qt/guiutil.cpp]: Add a comment that `UNICODE` is only defined for Windows? (if that's the case)

**Review by Sjors — COMMENTED** [review]
Should the reference also be removed from `test/lint/lint-includes.sh`?

Tested that `make` still works on macOS.

Tested the gitian binary on a Windows 10 VM, by launching with `bitcoin-qt -wallet=你好`, assuming that was the problem? This leads to a crash, but that's also the case on master:

schermafbeelding 2018-07-24 om 17 43 26

I have a version from earlier this year where it leads to a slightly nicer crash:
schermafbeelding 2018-07-24 om 17 46 37

(seems unrelated, so I made a ticket #13754)

`bitcoin-qt -wallet=test` works fine by the way, maybe that was all.

**Comment by ken2812221** [comment]
Update: Since the goal is to make it unicode compatible, just make it use wchar_t version of api explicitly.

**Comment by MarcoFalke** [comment]
Tested that for commit 6031f7f (master and this pull) I can launch regtest and testnet after re-login for a user with only ascii character in the name. (Does not work if the username is `a_ä_α_Б`)

**Comment by ken2812221** [comment]
@MarcoFalke Can you test this again? Now it should fix #13819.

**Comment by MarcoFalke** [comment]
Now it fails for commit d41914b (master and this pull) that it can not write to the directory:

screenshot from 2018-08-01 11-35-25

**Comment by ken2812221** [comment]
@MarcoFalke That should fixed by #13426. You can merge these two PR.

File: src/qt/guiutil.cpp [hunk 2]
```diff
@@ -547,52 +545,37 @@ bool SetStartOnSystemStartup(bool fAutoStart)
...
-            CLSCTX_INPROC_SERVER, IID_IShellLink,
+            CLSCTX_INPROC_SERVER, IID_IShellLinkW,
             reinterpret_cast<void**>(&psl));
 
         if (SUCCEEDED(hres))
         {
             // Get the current executable path
-            TCHAR pszExePath[MAX_PATH];
-            GetModuleFileName(nullptr, pszExePath, sizeof(pszExePath));
+            wchar_t pszExePath[MAX_PATH];
```
[the remark this answers is not in the document]
Comment (ken2812221) [reply src/qt/guiutil.cpp]: fixed

File: src/util.cpp [hunk 3]
```diff
@@ -1118,9 +1118,9 @@ void AllocateFileRange(FILE *file, unsigned int offset, unsigned int length) {
 #ifdef WIN32
 fs::path GetSpecialFolderPath(int nFolder, bool fCreate)
 {
-    char pszPath[MAX_PATH] = "";
+    wchar_t pszPath[MAX_PATH] = L"";
```
[the remark this answers is not in the document]
Comment (ken2812221) [reply src/util.cpp]: fixed

File: src/util.cpp [hunk 4]
```diff
@@ -1118,9 +1118,9 @@ void AllocateFileRange(FILE *file, unsigned int offset, unsigned int length) {
 #ifdef WIN32
 fs::path GetSpecialFolderPath(int nFolder, bool fCreate)
 {
-    char pszPath[MAX_PATH] = "";
+    WCHAR pszPath[MAX_PATH] = L"";
 
-    if(SHGetSpecialFolderPathA(nullptr, pszPath, nFolder, fCreate))
+    if(SHGetSpecialFolderPathW(nullptr, pszPath, nFolder, fCreate))
```
Comment (Empact) [inline src/util.cpp]: nit: the failure message below references `SHGetSpecialFolderPathA`

**Comment by Empact** [comment]
utACK b446918

File: src/util.cpp (hunk 4 above)
Comment (ken2812221) [reply src/util.cpp]: fixed

**Comment by laanwj** [comment]
nice cleanup, too

utACK e75b65c425138110bec72dc05e65dd9f4ca478bf

**Pull request merged**